### PR TITLE
[clang-tidy] Add modernize-use-std-format check

### PR DESCRIPTION
**Pull request by mikecrowe**
Add a new clang-tidy check that converts absl::StrFormat (and similar functions) to std::format (and similar functions.)

Split the configuration of FormatStringConverter out to a separate Configuration class so that we don't risk confusion by passing two boolean configuration parameters into the constructor. Add AllowTrailingNewlineRemoval option since we never want to remove trailing newlines in this check.

### CODE REVIEW

**Comment by mikecrowe** [comment]
This check was originally proposed and reviewed at https://reviews.llvm.org/D154287 last year but never landed since I did not (and still do not) have commit access. Near the end of that review, @PiotrZSL wrote:
> To be honest, I do not see to much use case for this check. But that's fine.

which may mean that it should not land. I've created this pull request for discoverability by those looking for such a check since Phabricator is no longer used.

File: clang-tools-extra/clang-tidy/modernize/UseStdFormatCheck.h
Comment (EugeneZelenko) [inline clang-tools-extra/clang-tidy/modernize/UseStdFormatCheck.h:12]: Included twice.

File: clang-tools-extra/clang-tidy/modernize/UseStdFormatCheck.cpp
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/modernize/UseStdFormatCheck.cpp:94]: `<< OldFunction; ` should be sufficient.

File: clang-tools-extra/clang-tidy/modernize/UseStdFormatCheck.cpp
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/modernize/UseStdFormatCheck.cpp:96]: getBeginLoc/getEndLoc could be changed into [getSourceRange](https://clang.llvm.org/doxygen/classclang_1_1Stmt.html#a6844392ee6148b5fe821f211b95e5d1b) ()

File: clang-tools-extra/clang-tidy/utils/FormatStringConverter.cpp
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/utils/FormatStringConverter.cpp:634]: Avoid constructing StringRef multiple times.

File: clang-tools-extra/clang-tidy/modernize/UseStdFormatCheck.h
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/modernize/UseStdFormatCheck.h:17]: synchronize with release notes

File: clang-tools-extra/test/clang-tidy/checkers/modernize/use-std-format.cpp
Comment (PiotrZSL) [inline clang-tools-extra/test/clang-tidy/checkers/modernize/use-std-format.cpp:97]: add few tests with macros. Mainly when part of call is behind macro (name, arguments).

File: clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-format.rst
Comment (PiotrZSL) [inline clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-format.rst:56]: `.)` -> `).`

**Review by PiotrZSL — APPROVED** [review]
To be honest I'm not sure why it wasn't merged, there were some dependences on other changes and modernize-use-std-print check, and somehow it fall under radar.

Overall looks fine. Please fix pointed out nits that pop up.

File: clang-tools-extra/test/clang-tidy/checkers/modernize/use-std-format.cpp
Comment (mikecrowe) [reply clang-tools-extra/test/clang-tidy/checkers/modernize/use-std-format.cpp:97]: I've added some tests. The check as currently implemented sees through and replaces the macro for the function call and argument but does not replace macros in the argument list.

File: clang-tools-extra/clang-tidy/modernize/UseStdFormatCheck.cpp
Comment (SimplyDanny) [inline clang-tools-extra/clang-tidy/modernize/UseStdFormatCheck.cpp:85]:
```suggestion
                                  "unable to use '%0' instead of '%1' because %2")
```

File: clang-tools-extra/clang-tidy/modernize/UseStdFormatCheck.cpp
Comment (SimplyDanny) [inline clang-tools-extra/clang-tidy/modernize/UseStdFormatCheck.cpp:93]:
```suggestion
      diag(StrFormatCall->getBeginLoc(), "use '%0' instead of '%1'")
```

File: clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-format.rst
Comment (SimplyDanny) [inline clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-format.rst:12]:
```suggestion
For example, it turns lines like
```

File: clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-format.rst
Comment (SimplyDanny) [inline clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-format.rst:39]:
```suggestion
   `StrictMode` enabled,
```

File: clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-format.rst
Comment (SimplyDanny) [inline clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-format.rst:48]:
```suggestion
  would be converted to
```

**Review by SimplyDanny — APPROVED** [review]
Only a few nits concerning non-functional code/text. Other than that LGTM.

File: clang-tools-extra/clang-tidy/modernize/UseStdFormatCheck.cpp
Comment (mikecrowe) [reply clang-tools-extra/clang-tidy/modernize/UseStdFormatCheck.cpp:85]: The use of `OldFunction->getIdentifier()` means that the string passed for `%1` is already surrounded by quotes. You can see this in the lit test strings:
```
// CHECK-MESSAGES: [[@LINE-1]]:13: warning: use 'std::format' instead of 'StrFormat' [modernize-use-std-format]
```

Comment (mikecrowe) [reply clang-tools-extra/clang-tidy/modernize/UseStdFormatCheck.cpp:85]: Yes. I can't say that I particularly like the situation and if I should be using a different method to get the function name then I'd be happy to change it. If someone changes that behaviour in the future then the lit tests will all start failing so they will be alerted to the need to do something.

File: clang-tools-extra/test/clang-tidy/checkers/modernize/use-std-format.cpp
Comment (mikecrowe) [reply clang-tools-extra/test/clang-tidy/checkers/modernize/use-std-format.cpp:97]: https://github.com/llvm/llvm-project/blob/84a81343cdda86efa34a9340f2a7ab7d2b388731/clang-tools-extra/test/clang-tidy/checkers/modernize/use-std-format.cpp#L99

@PiotrZSL, were these the kinds of tests you were expecting when you made your request and are you happy with the behaviour of the check for them? If not, please can you provide more specific advice?

**Review by PiotrZSL — APPROVED** [review]
LGTM

File: clang-tools-extra/test/clang-tidy/checkers/modernize/use-std-format.cpp
Comment (PiotrZSL) [inline clang-tools-extra/test/clang-tidy/checkers/modernize/use-std-format.cpp:104]: but, this shound't be fixed, so for macros like this there should be no fixes

File: clang-tools-extra/test/clang-tidy/checkers/modernize/use-std-format.cpp
Comment (PiotrZSL) [inline clang-tools-extra/test/clang-tidy/checkers/modernize/use-std-format.cpp:111]: no fix also should be here

File: clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-format.rst
Comment (mikecrowe) [inline clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-format.rst:16]: The newline in the format string here should be removed since it isn't present in the format string on line 22 below.

File: clang-tools-extra/test/clang-tidy/checkers/modernize/use-std-format.cpp
Comment (mikecrowe) [reply clang-tools-extra/test/clang-tidy/checkers/modernize/use-std-format.cpp:104]: Would you be willing to let this change land as it is and then I'll work on fixing this problem in a subsequent change for both this check and for _modernise-use-std-print_ that suffers from the same problem?

File: clang-tools-extra/test/clang-tidy/checkers/modernize/use-std-format.cpp
Comment (mikecrowe) [reply clang-tools-extra/test/clang-tidy/checkers/modernize/use-std-format.cpp:111]: I think this is required behaviour otherwise the check wouldn't be able to convert the unfortunately normal:
```C++
std::string f(int64_t i, uint64_t u) {
  return strprintf("%" PRId64 ", %" PRIu64, i, u);
}
```
to the expected
```C++
std::string f(int64_t i, uint64_t u) {
  return std::format("{}, {}", i, u);
}
```
The _modernise-use-std-print_ check currently does the same thing and is [documented](https://clang.llvm.org/extra/clang-tidy/checks/modernize/use-std-print.html) as doing so.

File: clang-tools-extra/test/clang-tidy/checkers/modernize/use-std-format.cpp
Comment (PiotrZSL) [reply clang-tools-extra/test/clang-tidy/checkers/modernize/use-std-format.cpp:104]: Thing is, that macro can be configuration dependent, and removing macros is not a good idea.
Add also test where everything is under macro (if there arent anny).

Once build will pass, I will land this change, and fixes could be done in separate PR for both checks.

File: clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-format.rst
Comment (5chmidti) [inline clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-format.rst:64]: The `extra` is wrong here. If you don't pass an option the default is described later, but when you do pass something, then default `absl::StrFormat` is not part of the checked function names.

File: clang-tools-extra/test/clang-tidy/checkers/modernize/use-std-format-custom.cpp
Comment (5chmidti) [inline clang-tools-extra/test/clang-tidy/checkers/modernize/use-std-format-custom.cpp:1]: Should the check options use the simpler format? E.g. 
```
CheckOptions: 
  modernize-use-std-format.StrFormatLikeFunctions: 'unqualified_strprintf;::strprintf; mynamespace::strprintf2'
```

**Review by 5chmidti — APPROVED** [review]
Just nits, otherwise LGTM!

File: clang-tools-extra/clang-tidy/modernize/UseStdFormatCheck.cpp
Comment (mikecrowe) [reply clang-tools-extra/clang-tidy/modernize/UseStdFormatCheck.cpp:84]: True. I'll fix that.

File: clang-tools-extra/clang-tidy/modernize/UseStdFormatCheck.cpp
Comment (mikecrowe) [reply clang-tools-extra/clang-tidy/modernize/UseStdFormatCheck.cpp:88]: OK. It looks like use-std-print needs that fix too, so I'll do that in a separate change. Is there a way to get the lit test to check for this?

File: clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-format.rst
Comment (mikecrowe) [reply clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-format.rst:64]: True. I'll fix that.

File: clang-tools-extra/test/clang-tidy/checkers/modernize/use-std-format-custom.cpp
Comment (mikecrowe) [reply clang-tools-extra/test/clang-tidy/checkers/modernize/use-std-format-custom.cpp:1]: Thanks! I wasn't aware that was possible since I started with all these checks before fbf611ed2a768999202e2c5e1e1a6c3c6bb94725 and had been battling with the JSON syntax.

File: clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-format.rst
Comment (mikecrowe) [reply clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-format.rst]: The fix for use-std-print is in https://github.com/llvm/llvm-project/pull/91069 .

File: clang-tools-extra/clang-tidy/modernize/UseStdFormatCheck.cpp
Comment (mikecrowe) [reply clang-tools-extra/clang-tidy/modernize/UseStdFormatCheck.cpp]: The use-std-print equivalent is in https://github.com/llvm/llvm-project/pull/91071 .

**Comment by mikecrowe** [comment]
@PiotrZSL wrote:
> Once build will pass, I will land this change, and fixes could be done in separate PR for both checks.

The build seems to be passing. Please can you land this if you are happy?

Thanks.

**Comment by kadircet** [comment]
hi! this seems to be crashing for certain code patterns, https://github.com/llvm/llvm-project/issues/92896 has a minimal reproducer

**Pull request merged**
